src/cmd/sh.go

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright © 2019 – 2021 Red Hat Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"os/user"
+	"strings"
+	"syscall"
+
+	"github.com/containers/toolbox/pkg/utils"
+	"github.com/spf13/cobra"
+)
+
+var shCmd = &cobra.Command{
+	Use:    "sh",
+	Short:  "Launch user configured shell inside container",
+	Hidden: true,
+	RunE:   sh,
+}
+
+func init() {
+	shCmd.SetHelpFunc(shHelp)
+	rootCmd.AddCommand(shCmd)
+}
+
+func sh(cmd *cobra.Command, args []string) error {
+	if !utils.IsInsideContainer() {
+		var builder strings.Builder
+		fmt.Fprintf(&builder, "the 'sh' command can only be used inside containers\n")
+		fmt.Fprintf(&builder, "Run '%s --help' for usage.", executableBase)
+
+		errMsg := builder.String()
+		return errors.New(errMsg)
+	}
+
+	u, err := user.Current()
+	if err != nil {
+		return fmt.Errorf("failed to get current user: %w", err)
+	}
+
+	out, err := exec.Command("getent", "passwd", u.Username).Output()
+	if err != nil {
+		if err, ok := err.(*exec.ExitError); ok {
+			return fmt.Errorf("failed to get user shell: %w", err)
+		}
+		return err
+	}
+
+	s := strings.Split(strings.TrimSuffix(string(out), "\n"), ":")[6]
+
+	err = syscall.Exec(s, []string{"-l"}, os.Environ())
```

Comment:
We will eventually need a fallback mechanism here.

Even today, if the user's shell on the *host* is Z shell, then right after `toolbox create`, the container is unlikely to have `/bin/zsh` in it.  So, `toolbox enter` will show an error message, but still fall back to `/bin/bash`, to let the user enter the container and install Z shell there.  From there on, `toolbox enter` will keep using `/bin/zsh`.

It will be good to eventually fold that into this new command, so that the fallback mechanism doesn't need to go through several invocations of `podman exec`, which is slow.

Since, we are now also looking at the user's default shell *inside* the container, there's a possibility that the shell is not available inside the container.  I agree that this is unlikely, but it's possible that the shell got uninstalled because of a mistake.  If we don't offer a fallback in this case, then it will be quite cumbersome for the user to recover the container.  The default shell on the host could be a fallback option.

Note that `toolbox enter` is meant to be used interactively, so it's different from something like [nologin(8)](https://man7.org/linux/man-pages/man8/nologin.8.html).

Anyway, doing all that will require moving around a good amount of code, so it's better to keep it simple in the first commit, and build upon it subsequently.  However, to not break `git bisect`, we need to ensure that even the initial simple implementation uses an exit code of 127 to keep the existing (slow) fallback working.

Reply:
I’m thinking about the way to implement the fallback mechanism, it’s not ready at the time.

Reply:
I think I’ve a working fallback mechanism.